README.md
#### todo

- move all the db config into local external files not checked into the repo, which are read once when
connecting to the db.

- find out file names which git will always ignore because they (nearly) always contain sensitive info

#### hugsql-demo

Install leiningen. Clone this repo. 

You'll need Postgres or SQLite. If you're running Postgres, the db server must be started. Modify the dbspec
in core.clj for your database.

Setting a Postgres password will no have any effect until you change local connections in pg_hba.conf from
"trust" to "md5". Postgres trusted connections do not require a password, but are only safe for connections on
the localhost (and trust is only valid for localhost).

Run your database command line, and load the appropriate schema file. 

For postgres it might be simpler to use the command line `createuser` and `createdb`, but the schema file seems to work.

```
> psql postgres
postgres=# \i resources/schema.sql
postgres=# \q
> 


> sqlite3 hdemo.db
sqlite> .read resources/schema_sqlite.sql
sqlite> .q
>

```

Put your Postgres database password in the local file .dbpass.txt and do not check this file into version
control. (Yes the file name begins with a dot and is therefore called a "dotfile". Dot files are often
configuration, and often not checked into version control.)

```
> psql hdemo
postgres=# alter user dbuser with encrypted password 'foobar';
postgres=# \q
> echo "foobar" > .dbpass.txt
```


Launch the repl and run the demo functions from the repl. If you aren't running cider, you can run
(tnr/refresh) in the repl after making changes to the code, in order to force the repl to pick up the changes.

```
hugsql-demo.core=> (tnr/refresh)
:reloading (hugsql-demo.core hugsql-demo.core-test)
:ok
hugsql-demo.core=> (time (demo-vector dbspec-sqlite))
"Elapsed time: 152.925096 msecs"
10000
hugsql-demo.core=> (time (demo-vector dbspec-sqlite))
"Elapsed time: 164.23081 msecs"
10000
hugsql-demo.core=> (time (demo-vector dbspec))
"Elapsed time: 677.292442 msecs"
10000
hugsql-demo.core=> (time (demo-vector dbspec))
"Elapsed time: 626.836757 msecs"
10000
hugsql-demo.core=> 
```

#### Links to docs

http://funcool.github.io/clojure.jdbc/latest/

https://www.hugsql.org/



#### fetch and execute are not interchangeable

Anyone who knows me knows that I'm fond of Perl DBD. There's a long list of great features in DBD, many of
which are missing from other database connectivity libraries. I'm convinced that Tim Bunce is a genius. One
example is that DBD execute simply works on everything. This is not true in Clojure. However, if you are lucky
enough to be using Postgres and if you want to always use fetch, then "returning ..." is your friend.

Certainly, it doesn't take much thinking to know if you need fetch or execute. Still, why is the high level
programmer being bothered by this kind of trivia? If I want to waste the output of my mitochondria on trivia
I'll program in Go or C.

I haven't tried this in real life, but "returning id" works (or "returning 1" kind of, if only 1 row is
effected). For an accurate count of rows affected, one would have to write a plpgsql function, which is far
too much hassle.

https://stackoverflow.com/questions/38875955/postgresql-function-return-affected-row-count

```
hugsql-demo.core=> (def conn (jdbc/connection dbspec))
#'hugsql-demo.core/conn
hugsql-demo.core=> (jdbc/fetch conn ["select 1"])
[{:?column? 1}]
hugsql-demo.core=> (jdbc/execute conn ["select 1"])

PSQLException A result was returned when none was expected.  org.postgresql.jdbc.PgPreparedStatement.executeUpdate (PgPreparedStatement.java:141)
hugsql-demo.core=> (jdbc/fetch conn ["insert into address (city) values ('foo')"])

PSQLException No results were returned by the query.  org.postgresql.jdbc.PgPreparedStatement.executeQuery (PgPreparedStatement.java:118)
hugsql-demo.core=> (jdbc/fetch conn ["insert into address (city) values ('foo') returning id"])
[{:id 347018}]
hugsql-demo.core=> (jdbc/fetch conn ["insert into address (city) values ('foo') returning 1"])
[{:?column? 1}]
hugsql-demo.core=> (jdbc/fetch conn ["update address set city='bar' where city='foo' returning id;"])
[{:id 347017} {:id 347018} {:id 347019}]

```

Really, just use execute for insert, update, delete. Use fetch for select.



#### Prepared statements

Prepare is not supported by HugSQL, nor explicitly by clojure.jdbc (except for internal use in fetch-lazy). 

Clojure.java.jdbc has prepareed statements. See core.clj functions demo-java-jdbc-t1 and demo-java-jdbc-t2.

However:

- prepared statements are awkward for several reasons. prepare-statement takes a raw connection, but
  with-db-transaction, execute! and query all take a {:connection conn}

- clojure.jdbc is faster running repeated parameterized SQL inside a transaction than a clojure.java.jdbc
  prepared statement inside a transaction

- clojure.java.jdbc requires more lines of code


#### todo

* There's no error checking. Need to deal with that because silent SQL failures can be really irritating bugs.

* Might be a good idea to include working clojure.java.jdbc code.

* Might be nice to turn this into a real app. It won't ever do much, but right now everything interesting is
  limited to the repl, so this is hardly a completely working example.

* Need to test and/or demo cursors. Postgres has them, but we are better off using some level of abstraction.



#### HugSQL header overview

Shorthand and longhand versions of a HugSQL header in a sql file. See resources/full.sql.

The :command specifies what kind of action. The :result specifies the result data structure. We will call a
clojure.lang.PersistentArrayMap a "hash-map".

```
-- A comment. Assume id is a unique primary key.
-- :name foo :? :1
-- :doc Query the database returning a single record.
select * from foo where id = :id

-- A comment. Assume id is a unique primary key.
-- :name foo
-- :command :query
-- :result :one
-- :doc Query the database returning a single record.
select * from foo where id = :id

```

Available :command values:

- :query or :? = query with a result-set (default)
- :execute or :! = any statement
- :returning-execute or :<! = support for INSERT ... RETURNING
- :insert or :i! = support for insert and jdbc .getGeneratedKeys

:query and :execute mirror the distinction between query and execute! in the clojure.java.jdbc library and
fetch and execute in the clojure.jdbc library.


| command  | clojure.java.jdbc | clojure.jdbc |
|----------|-------------------|--------------|
| :query   | query             | fetch        |
| :execute | execute!          | execute      |


Available :result values:

- :one or :1 = one row, a hash-map
- :many or :* = many rows, a vector of hash-map
- :affected or :n = number of rows affected (inserted/updated/deleted), an integer
- :raw = passthrough an untouched result (default)

The default :result is :raw, which may be omitted. The default seems to be a vector of hash-map, but you
should verify this for your database connection library.



#### Odd errors you may encounter

If you leave the db connection arg off a SQL function, you will get a less-than-obvious error:

```
ExceptionInfo Parameter Mismatch: :street parameter data not found.  clojure.core/ex-info (core.clj:4617)
```

Incorrect hash-map of vector passed as :tuple*

```
PSQLException Can't infer the SQL type to use for an instance of clojure.lang.PersistentVector. Use setObject() with an explicit Types value to specify the type to use.  org.postgresql.jdbc.PgPreparedStatement.setObject (PgPreparedStatement.java:973)
```


This is caused by having a bad db-spec:

```
NoInitialContextException Need to specify class name in environment or system property, or as an applet parameter, or in an application resource file:  java.naming.factory.initial  javax.naming.spi.NamingManager.getInitialContext (NamingManager.java:662)
```

Correct docs here:

https://github.com/clojure-cookbook/clojure-cookbook/blob/master/06_databases/6-01_connecting-to-an-SQL-database.asciidoc

Wrong or out of date docs here:

http://peterstratton.com/posts-output/2017-01-28-postgres-and-clojure-using-clojure-java-jdbc/


```
(def dbspec-java-jdbc
  {:classname   "org.postgresql.Driver"
   :subprotocol "postgresql"
   :subname     "hdemo"
   :user "dbuser"
   :password (pw-read)})

(def dbspec-java-example
  {:classname "org.postgresql.Driver"
   :subprotocol "postgresql"
   :subname "//localhost:5432/hdemo" ;; is this //<host>:<port>/<dbname> ??
   :user "dbuser"
   :password "foobar"})

;; Wrong
(def bad-dbspec-java-jdbc
  {:dbtype "postgresql"
   :name "hdemo"
   :host "localhost"      ;; Optional
   :port 5432             ;; Optional
   :user "deuser"
   :password "foobar"
   :ssl false
   :sslfactory "org.postgresql.ssl.NonValidatingFactory"})
```

Fix this by changing the database to be owned by the user `alter database hdemo owner to dbuser;`

```
PSQLException ERROR: permission denied for relation address  org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse (QueryExecutorImpl.java:2455)
```


## License

Copyright © 2017 Tom Laudeman

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-12-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-12-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hugsql-demo/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hugsql-demo/compare/0.1.0...0.1.1


test/hugsql_demo/core_test.clj
(ns hugsql-demo.core-test
  (:require [clojure.test :refer :all]
            [hugsql-demo.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject hugsql-demo "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [com.layerware/hugsql "0.4.8"] ;; use with clojure.java.jdbc
                 [com.layerware/hugsql-core "0.4.8"] ;; use with clojure.jdbc
                 [com.layerware/hugsql-adapter-clojure-jdbc "0.4.8"] ;; use with clojure.jdbc
                 ;; https://github.com/funcool/clojure.jdbc
                 [funcool/clojure.jdbc "0.9.0"] ;; use with hugsql-core and hugsql-adapter-clojure-jdbc
                 [org.clojure/java.jdbc "0.7.5"]
                 [org.clojure/tools.namespace "0.2.11"]
                 [org.xerial/sqlite-jdbc "3.21.0"]
                 [org.postgresql/postgresql "42.1.4"]
                 [org.postgresql/postgresql "9.4.1212"]]

  ;; The tutorial may be out of date. the :uberjar profile probably obviates :main and :aot.
  ;; https://github.com/technomancy/leiningen/blob/master/doc/TUTORIAL.md#uberjar
  ;; :main my-stuff.core
  ;; :aot [my-stuff.core]

  :main ^:skip-aot hugsql-demo.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


src/hugsql_demo/core.clj
(ns hugsql-demo.core
  (:require [clojure.tools.namespace.repl :as tnr]
            [clojure.repl :refer [doc]]
            [clojure.java.shell]
            [clojure.java.io]
            [jdbc.core :as jdbc] ;; clojure.jdbc
            [clojure.java.jdbc]
            [clojure.string :as str]
            [hugsql.core :as hugsql]
            [hugsql.adapter.clojure-jdbc :as cj-adapter])
  (:gen-class))

(defn pw-read []
  "Read the password file and return the password. Return an empty string on any error."
  (str/trim
   (try (slurp ".dbpass.txt")
        (catch Exception e (do 
                             (.printStackTrace e)
                             "")))))

(def dbspec-sqlite
  {:classname   "org.sqlite.JDBC"
   :subprotocol "sqlite"
   :subname     "hdemo.db"
   })

;; https://github.com/clojure/java.jdbc
;; This is not correct, in spite of what the docs seem to say
(def dbspec
  {:dbtype "postgresql"
   :classname   "org.postgresql.Driver"
   :subprotocol "postgresql"
   :dbname "hdemo" ;; db name
   :subname "//localhost:5432/hdemo"
   :user "dbuser"
   :password (pw-read)
   :port 5432
   :hostname "127.0.0.1"})

;; Only if you have enabled ssl on your postgres server.
;; :ssl true
;; :sslfactory "org.postgresql.ssl.NonValidatingFactory"

;; https://github.com/clojure-cookbook/clojure-cookbook/blob/master/06_databases/6-01_connecting-to-an-SQL-database.asciidoc
;; This may only work with the old, previous generation postgres driver [postgresql/postgresql "9.3-1102.jdbc41"]
(def dbspec-java-jdbc
  {:classname   "org.postgresql.Driver"
   :subprotocol "postgresql"
   :subname     "hdemo"
   :user "dbuser"
   :password (pw-read)})

;; Hugsql macros must be outside defn and come before any mention of functions that they will create at
;; runtime. Two functions will be created for each :name in full.sql. 

(hugsql/def-db-fns "hugsql_demo/full.sql"
  {:adapter (cj-adapter/hugsql-adapter-clojure-jdbc)})

(hugsql/def-sqlvec-fns "hugsql_demo/full.sql"
  {:adapter (cj-adapter/hugsql-adapter-clojure-jdbc)})


;; 2570 ms/10K records, essentially the same as the far simpler (demo-transaction) below.
(defn demo-java-jdbc-t2
  "Prepare a statement with placeholder params, and run inside a transaction."
  []
  (let [conn (clojure.java.jdbc/get-connection dbspec-java-jdbc)
        stmt (clojure.java.jdbc/prepare-statement conn "insert into address (street, city, postal_code) values (?, ?, ?)")]
    (clojure.java.jdbc/with-db-transaction [tx {:connection conn}]
      (doseq [xx (range 10000)]
        (clojure.java.jdbc/execute!
         tx [stmt "Bayview" (str xx " Maple Street") "93312"])))))


;; 3100 ms/10K records. Roughtly 30% slower than clojure.jdbc, and slower than a non-prepared statement.
(defn demo-java-jdbc-t1
  "Run inserts inside a transaction. Sometimes a connection is a {:connection conn}. I can't decipher the
  difference in the docs."
  []
  (let [conn {:connection (clojure.java.jdbc/get-connection dbspec-java-jdbc)}]
    (clojure.java.jdbc/with-db-transaction [tx conn]
      (doseq [xx (range 10000)]
        (clojure.java.jdbc/execute!
         tx
         (insert-address-sqlvec {:city "Bayview" :street (str xx " Maple Street") :postal_code "93312"}))))))


;; The HugSQL docs point out that tuple list param only works for insert, it depends support from the
;; underlying db, and there is a limit to the size of the param vector. I would add that if you want raw
;; insert speed, you need to be using the COPY command. Still, this is cool.

;; 650 ms/10K
;; That is 3x better than transactions, 6x better than auto-commiting each insert as a separate transaction.
;; I need another demo that uses a prepared statement, which will should shave off a little more time.
(defn demo-vector
  [dbs]
  (let [conn (jdbc/connection dbs)]
    (insert-address-vector conn
                           {:address
                            (mapv (fn [xx] ["Bayview" (str xx " Maple Street") "93312"])
                                   (range 10000))})))


(defn demo-prep
  "This does not work. Prepared statements are not supported by clojure.jdbc, except that they are used
  internally for fetch-lazy using a cursor. Postgres supports prepared statements. Older code (Perl DBI)
  supports prepare for all relevant query types, so this feature is hardly new. I need to check support in
  clojure.java.jdbc."
  []
  (let [conn (jdbc/connection dbspec)]  
    (jdbc/atomic conn
                 (let [stmt (jdbc/prepared-statement conn ["insert into address (street, city, postal_code)\nvalues (?, ?, ?)"])]
                   (doseq [xx (range 10000)]
                     (jdbc/execute stmt [(str xx " Maple Street") "Bayview" "93312"]))))))
                                   
    
;; 210 ms/1k
;; 1970 ms/10k
;; After adding an index: 2250 md/10k
(defn demo-transaction
  []
  (let [conn (jdbc/connection dbspec)]
    (jdbc/atomic conn 
                 (doseq [xx (range 10000)]
                   (insert-address conn {:city "Bayview" :street (str xx " Maple Street") :postal_code "93312"})))))


;; 450 ms/1k
;; 4400 ms/10K
;; After adding an index: 4780 ms/10k
(defn demo-non-transaction
  []
  (let [conn (jdbc/connection dbspec)]
    (doseq [xx (range 10000)]
      (insert-address conn {:city "Bayview" :street (str xx " Maple Street") :postal_code "93312"}))))
  

(defn demo
  "Using clojure.jdbc. Run some code and def vars that you can inspect in the repl."
  []
  (def conn (jdbc/connection dbspec))
  (def check-res (jdbc/fetch conn ["SELECT 1"]))
  (jdbc/execute conn ["insert into address (city) values ('foo')"])
  (def in-res (insert-address conn {:city "Bayview" :street "110 Maple Street" :postal_code "93312"}))
  (def foo-res (address-ilike-city conn {:city "foo"}))
  (def bv-res (address-ilike-city conn {:city "bayview"})))

(defn -main
  "Using clojure.jdbc"
  [& args]
  (def conn (jdbc/connection dbspec))
  (def results (jdbc/fetch conn ["SELECT 1"])) ;; Verify the connection to the db.
  (jdbc/execute conn ["insert into address (city) values ('foo')"])
  (def aselect (address-ilike-city conn {:city "foo"})))

